pkg/sysfs/system.go

```diff
@@ -614,12 +628,39 @@ func (c *cpu) SetFrequencyLimits(min, max uint64) error {
 	return nil
 }
 
+func readCPUsetFile(base, entry string) (cpuset.CPUSet, error) {
+	path := filepath.Join(base, entry)
+
+	blob, err := ioutil.ReadFile(path)
+	if err != nil {
+		return cpuset.NewCPUSet(), sysfsError(path, "failed to read sysfs entry: %v", err)
+	}
+
+	return cpuset.Parse(strings.Trim(string(blob), "\n"))
+}
+
 // Discover NUMA nodes present in the system.
 func (sys *system) discoverNodes() error {
 	if sys.nodes != nil {
 		return nil
 	}
 
+	// FIXME assumption: if a node only has memory (and no CPUs), it's PMEM. Otherwise it's DRAM.
```

Comment:
we need to check size. If $amount is < than DRAM, then it is HBMEM. if > DRAM then PMEM.

Reply:
That's another assumption, which has it's own problems...

A proper way to do this would be to go through the ACPI tree and parse the entries there. or use existing tooling which does the same. I have the feeling that traversing the ACPI tree approach is also tricky, because we don't know beforehand the ACPI tree layout of each and every system, and testing it widely enough is hard. Using existing tooling would give us an  external dependency and possibly some headache with parsing tool output data and maintaining compatibility between versions.

I wrote the test below to find the PMEM nodes (and the NUMA nodes which map to them. It gives some idea of the potential complexity. I don't know if the ACPI tree files and their data layout is under the kernel<->userspace API stability promise...

```go
package main

import (
        "fmt"
        "io/ioutil"
        "os"
        "path"
        "path/filepath"
        "strconv"
        "strings"
)

const (
        acpiRoot = "/sys/devices/LNXSYSTM\\:00/"
)

type pmemregion struct {
        numaNode   int
        targetNode int
}

func main() {

        nodes := make([]pmemregion, 0)

        buses, _ := filepath.Glob(path.Join(acpiRoot, "LNXSYBUS\\:*"))
        for _, bus := range buses {
                devices, _ := filepath.Glob(path.Join(bus, "ACPI*"))
                for _, device := range devices {
                        ndbuses, _ := filepath.Glob(path.Join(device, "ndbus*"))
                        for _, ndbus := range ndbuses {
                                regions, _ := filepath.Glob(path.Join(ndbus, "region*"))
                                for _, region := range regions {
                                        buf, err := ioutil.ReadFile(path.Join(region, "numa_node"))
                                        if err != nil {
                                                fmt.Println("Error reading file")
                                                os.Exit(1)
                                        }
                                        numaNode, _ := strconv.ParseInt(strings.Trim(string(buf), "\n"), 0, 64)

                                        buf, err = ioutil.ReadFile(path.Join(region, "target_node"))
                                        if err != nil {
                                                fmt.Println("Error reading file")
                                                os.Exit(1)
                                        }

                                        targetNode, _ := strconv.ParseInt(strings.Trim(string(buf), "\n"), 0, 64)

                                        entry := pmemregion{numaNode: int(numaNode), targetNode: int(targetNode)}

                                        nodes = append(nodes, entry)
                                }
                        }
                }
        }

        for _, node := range nodes {
                fmt.Printf("Numa node: %d -> PMEM node %d\n", node.numaNode, node.targetNode)
        }
}
```